File: packages/claudewire/claudewire/transport.py
"""BridgeTransport -- SDK Transport that routes through a ProcessConnection.

Implements the claude_agent_sdk Transport interface (6 abstract methods).
For reconnecting agents, intercepts the initialize control_request and
fakes a success response -- the CLI is already initialized.

This module has NO dependency on procmux. It works with any backend that
satisfies the ProcessConnection protocol defined in claudewire.types.
"""

from __future__ import annotations

import json
import logging
from typing import TYPE_CHECKING, Any

from opentelemetry import trace
from opentelemetry.propagate import inject

from claudewire.types import CommandResult, ExitEvent, StderrEvent, StdoutEvent

if TYPE_CHECKING:
    from collections.abc import Callable

    from claudewire.types import ProcessConnection, ProcessEventQueue

log = logging.getLogger(__name__)
_tracer = trace.get_tracer(__name__)


class BridgeTransport:
    """SDK Transport that routes through a ProcessConnection for one agent.

    Implements the claude_agent_sdk.Transport interface (6 abstract methods).
    For reconnecting agents, intercepts the initialize control_request and
    fakes a success response -- the CLI is already initialized.
    """

    def __init__(
        self,
        name: str,
        conn: ProcessConnection,
        *,
        reconnecting: bool = False,
        stderr_callback: Callable[[str], None] | None = None,
        stdio_logger: logging.Logger | None = None,
    ):
        self._name = name
        self._conn = conn
        self._reconnecting = reconnecting
        self._stderr_callback = stderr_callback
        self._stdio_logger = stdio_logger
        self._queue: ProcessEventQueue | None = None
        self._ready = False
        self._cli_exited = False
        self._exit_code: int | None = None

    async def connect(self) -> None:
        """Register with the process connection for this agent's output."""
        _tracer.start_span("claudewire.connect", attributes={"agent.name": self._name}).end()
        self._queue = self._conn.register(self._name)
        self._ready = True

    async def spawn(self, cli_args: list[str], env: dict[str, str], cwd: str) -> CommandResult:
        """Tell the process connection to spawn a new CLI process."""
        _tracer.start_span("claudewire.spawn", attributes={"agent.name": self._name, "agent.cwd": cwd}).end()
        result = await self._conn.spawn(self._name, cli_args=cli_args, env=env, cwd=cwd)
        if not result.ok:
            raise RuntimeError(f"Spawn failed for '{self._name}': {result}")
        return result

    async def subscribe(self) -> CommandResult:
        """Subscribe to this agent's output (triggers buffer replay)."""
        result = await self._conn.subscribe(self._name)
        if not result.ok:
            raise RuntimeError(f"Subscribe failed for '{self._name}': {result}")
        return result

    async def write(self, data: str) -> None:
        """Write data to the CLI's stdin."""
        with _tracer.start_as_current_span(
            "claudewire.write",
            attributes={"agent.name": self._name, "data.bytes": len(data)},
        ):
            if not self._conn.is_alive:
                raise ConnectionError("Process connection is dead")
            msg = json.loads(data)

            # Intercept initialize for reconnecting agents -- fake success
            if (
                self._reconnecting
                and msg.get("type") == "control_request"
                and msg.get("request", {}).get("subtype") == "initialize"
            ):
                request_id = msg.get("request_id")
                fake_response: dict[str, Any] = {
                    "type": "control_response",
                    "response": {
                        "subtype": "success",
                        "request_id": request_id,
                        "response": {},
                    },
                }
                if self._queue:
                    await self._queue.put(StdoutEvent(name=self._name, data=fake_response))
                self._reconnecting = False
                return

            # Inject OTel trace context so downstream processes can link spans
            carrier: dict[str, str] = {}
            inject(carrier)
            if carrier:
                msg["_trace_context"] = carrier

            if self._stdio_logger:
                self._stdio_logger.debug(">>> STDIN  %s", json.dumps(msg))
            await self._conn.send_stdin(self._name, msg)

    async def read_messages(self):
        """Async generator yielding parsed JSON dicts from CLI stdout."""
        if not self._queue:
            return
        if not self._conn.is_alive:
            raise ConnectionError("Process connection is dead")
        while True:
            msg = await self._queue.get()
            if msg is None:
                raise ConnectionError("Process connection lost during read")
            if isinstance(msg, StdoutEvent):
                msg_type = msg.data.get("type", "?")
                log.debug("[read][%s] yielding stdout type=%s", self._name, msg_type)
                if self._stdio_logger:
                    self._stdio_logger.debug("<<< STDOUT %s", json.dumps(msg.data))
                yield msg.data
            elif isinstance(msg, StderrEvent):
                log.debug("[read][%s] stderr: %.200s", self._name, msg.text)
                if self._stdio_logger:
                    self._stdio_logger.debug("<<< STDERR %s", msg.text)
                if self._stderr_callback:
                    self._stderr_callback(msg.text)
            elif isinstance(msg, ExitEvent):
                log.debug("[read][%s] exit code=%s", self._name, msg.code)
                _tracer.start_span(
                    "claudewire.cli_exit",
                    attributes={"agent.name": self._name, "exit.code": msg.code or -1},
                ).end()
                if self._stdio_logger:
                    self._stdio_logger.debug("--- EXIT   code=%s", msg.code)
                self._cli_exited = True
                self._exit_code = msg.code
                return

    async def close(self) -> None:
        """Kill the CLI process and unregister."""
        _tracer.start_span("claudewire.close", attributes={"agent.name": self._name}).end()
        if self._ready and not self._cli_exited:
            try:
                await self._conn.kill(self._name)
            except Exception:
                pass
        self._conn.unregister(self._name)
        self._ready = False

    def is_ready(self) -> bool:
        return self._ready

    async def end_input(self) -> None:
        """Not needed for bridge mode -- CLI stdin stays open for multi-turn."""

    @property
    def cli_exited(self) -> bool:
        return self._cli_exited

File: axi-rs/crates/claudewire/src/transport.rs
//! BridgeTransport — routes messages through a process event queue.
//!
//! Implements the transport interface for Claude CLI communication.
//! For reconnecting agents, intercepts the initialize control_request and
//! fakes a success response — the CLI is already initialized.
//!
//! This module has NO dependency on procmux. It works with any backend that
//! sends ProcessEvents through a tokio mpsc channel.

use tokio::sync::mpsc;
use tracing::debug;

use crate::schema::is_bare_stream_type;
use crate::types::*;

/// Callback type for stderr output.
pub type StderrCallback = Box<dyn Fn(&str) + Send + Sync>;

/// Function to send stdin to a process.
pub type SendStdinFn =
    Box<dyn Fn(String, serde_json::Value) -> std::pin::Pin<Box<dyn std::future::Future<Output = anyhow::Result<()>> + Send>> + Send + Sync>;

/// Function to kill a process.
pub type KillFn =
    Box<dyn Fn(String) -> std::pin::Pin<Box<dyn std::future::Future<Output = anyhow::Result<()>> + Send>> + Send + Sync>;

/// BridgeTransport routes Claude CLI messages through a process connection.
pub struct BridgeTransport {
    name: String,
    reconnecting: bool,
    stderr_callback: Option<StderrCallback>,
    rx: Option<mpsc::UnboundedReceiver<ProcessEvent>>,
    tx: Option<mpsc::UnboundedSender<ProcessEvent>>,
    send_stdin: SendStdinFn,
    kill: KillFn,
    is_alive: Box<dyn Fn() -> bool + Send + Sync>,
    ready: bool,
    cli_exited: bool,
    exit_code: Option<i32>,
}

impl BridgeTransport {
    pub fn new(
        name: String,
        rx: mpsc::UnboundedReceiver<ProcessEvent>,
        tx: mpsc::UnboundedSender<ProcessEvent>,
        send_stdin: SendStdinFn,
        kill: KillFn,
        is_alive: Box<dyn Fn() -> bool + Send + Sync>,
        reconnecting: bool,
        stderr_callback: Option<StderrCallback>,
    ) -> Self {
        Self {
            name,
            reconnecting,
            stderr_callback,
            rx: Some(rx),
            tx: Some(tx),
            send_stdin,
            kill,
            is_alive,
            ready: true,
            cli_exited: false,
            exit_code: None,
        }
    }

    /// Write data to the CLI's stdin.
    pub async fn write(&mut self, data: &str) -> anyhow::Result<()> {
        if !(self.is_alive)() {
            anyhow::bail!("Process connection is dead");
        }

        let msg: serde_json::Value = serde_json::from_str(data)?;

        // Intercept initialize for reconnecting agents — fake success
        if self.reconnecting {
            if let Some("control_request") = msg.get("type").and_then(|v| v.as_str()) {
                if let Some("initialize") = msg
                    .get("request")
                    .and_then(|r| r.get("subtype"))
                    .and_then(|v| v.as_str())
                {
                    let request_id = msg
                        .get("request_id")
                        .and_then(|v| v.as_str())
                        .unwrap_or("");
                    let fake_response = serde_json::json!({
                        "type": "control_response",
                        "response": {
                            "subtype": "success",
                            "request_id": request_id,
                            "response": {},
                        },
                    });
                    if let Some(tx) = &self.tx {
                        tx.send(ProcessEvent::Stdout(StdoutEvent {
                            name: self.name.clone(),
                            data: fake_response,
                        }))
                        .ok();
                    }
                    self.reconnecting = false;
                    return Ok(());
                }
            }
        }

        // Inject OTel trace context (placeholder — will wire up later)
        // For now, just send the message as-is
        (self.send_stdin)(self.name.clone(), msg).await
    }

    /// Async iterator yielding parsed JSON dicts from CLI stdout.
    /// Returns None when the stream ends.
    pub async fn read_message(&mut self) -> Option<serde_json::Value> {
        let rx = self.rx.as_mut()?;
        if !(self.is_alive)() && !self.cli_exited {
            return None;
        }

        loop {
            let event = rx.recv().await?;

            // stop() was called — discard buffered messages
            if self.cli_exited {
                return None;
            }

            match event {
                ProcessEvent::Stdout(stdout) => {
                    let msg_type = stdout
                        .data
                        .get("type")
                        .and_then(|v| v.as_str())
                        .unwrap_or("?");

                    // Drop bare duplicate stream events
                    if is_bare_stream_type(msg_type) {
                        debug!(
                            "[read][{}] dropping bare duplicate type={}",
                            self.name, msg_type
                        );
                        continue;
                    }

                    debug!("[read][{}] yielding stdout type={}", self.name, msg_type);
                    return Some(stdout.data);
                }
                ProcessEvent::Stderr(stderr) => {
                    debug!("[read][{}] stderr: {:.200}", self.name, stderr.text);
                    if let Some(cb) = &self.stderr_callback {
                        cb(&stderr.text);
                    }
                }
                ProcessEvent::Exit(exit) => {
                    debug!("[read][{}] exit code={:?}", self.name, exit.code);
                    self.cli_exited = true;
                    self.exit_code = exit.code;
                    return None;
                }
            }
        }
    }

    /// Immediately terminate the read stream and kill the CLI process.
    pub async fn stop(&mut self) {
        if self.cli_exited {
            return;
        }
        self.cli_exited = true;
        debug!(
            "[stop][{}] injecting ExitEvent and scheduling background kill",
            self.name
        );

        // Inject ExitEvent to unblock read_message()
        if let Some(tx) = &self.tx {
            tx.send(ProcessEvent::Exit(ExitEvent {
                name: self.name.clone(),
                code: None,
            }))
            .ok();
        }

        // Kill the process (best effort)
        let name = self.name.clone();
        if let Err(e) = (self.kill)(name).await {
            debug!(
                "[stop][{}] kill failed (process may already be dead): {}",
                self.name, e
            );
        }
    }

    /// Kill the CLI process and clean up.
    pub async fn close(&mut self) {
        if self.ready && !self.cli_exited {
            let name = self.name.clone();
            (self.kill)(name).await.ok();
        }
        self.ready = false;
    }

    pub fn is_ready(&self) -> bool {
        self.ready
    }

    pub fn cli_exited(&self) -> bool {
        self.cli_exited
    }

    pub fn exit_code(&self) -> Option<i32> {
        self.exit_code
    }

    pub fn name(&self) -> &str {
        &self.name
    }
}
